File: src/main/java/kitchenpos/products/application/ChangeProductPriceService.java
package kitchenpos.products.application;

import kitchenpos.products.tobe.domain.entity.Product;
import kitchenpos.products.tobe.domain.repository.ProductRepository;
import kitchenpos.common.domain.vo.Price;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.NoSuchElementException;
import java.util.UUID;

@Service
public class ChangeProductPriceService {
    private final ProductRepository productRepository;

    public ChangeProductPriceService(
        final ProductRepository productRepository
    ) {
        this.productRepository = productRepository;
    }

    @Transactional
    public Product changePrice(final UUID productId, final BigDecimal price) {
        final Price productPrice = new Price(price);
        final Product product = productRepository.findById(productId)
                .orElseThrow(NoSuchElementException::new);

        product.changePrice(productPrice);

        productRepository.save(product);

        return product;
    }
}

File: src/main/java/kitchenpos/products/tobe/application/ChangeProductPriceService.kt
package kitchenpos.products.tobe.application

import kitchenpos.products.domain.Product
import kitchenpos.products.domain.ProductRepository
import kitchenpos.products.tobe.domain.ProductPriceValidator
import kitchenpos.products.tobe.domain.UpdateProductPriceService
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.math.BigDecimal
import java.util.*

@Transactional
@Service
class ChangeProductPriceService(
    private val productRepository: ProductRepository,
) {
    fun changeProductPrice(
        productId: UUID,
        price: BigDecimal?,
    ): Product {
        ProductPriceValidator.requireNormalPrice(price)

        return productRepository.findById(productId).orElseThrow { NoSuchElementException() }
            .apply {
                UpdateProductPriceService.updatePrice(this, price!!)
            }
    }
}
